slack-api/src/main/java/pl/hellsoft/slack/wrapper/rtm/SlackRealTimeMessagingClient.kt

```diff
@@ -0,0 +1,195 @@
+ * @since 23/01/2018.
+ */
+class SlackRealTimeMessagingClient(val webSocketUrl: String?, val proxyServerInfo: ProxyServerInfo? = null, val pingMillis: Int = 3000) {
+
+    init {
+        d { "SlackRealTimeMessagingClient init" }
+    }
+    companion object {
+        const val NORMAL_CLOSURE_STATUS = 1000
+    }
+
+    private lateinit var disposables: CompositeDisposable
+
+    private var client: OkHttpClient? = null
+    private var ws: WebSocket? = null
+    private val gson by lazy { GsonBuilder().create() }
+
+
+    private val listeners = HashMap<String, MutableList<EventListener>>()
+    private val closeListeners = ArrayList<CloseListener>()
+    private val failureListeners = ArrayList<FailureListener>()
+
+    private var socketId: Long = 0
+
+    fun addListener(event: Event, listener: EventListener) {
+        addListener(event.name.toLowerCase(), listener)
+    }
+
+    private fun addListener(event: String, listener: EventListener) {
+        var eventListeners: MutableList<EventListener>? = listeners[event]
+        if (eventListeners == null) {
+            eventListeners = ArrayList()
+            listeners[event] = eventListeners
+        }
+        eventListeners.add(listener)
+    }
+
+    fun addCloseListener(listener: CloseListener) {
+        closeListeners.add(listener)
+    }
+
+    fun addFailureListener(listener: FailureListener) {
+        failureListeners.add(listener)
+    }
+
+    private val webSocketListener = object : WebSocketListener() {
+        override fun onOpen(webSocket: WebSocket, response: Response) {
+            i { "Connected to Slack RTM (Real Time Messaging) server : " + webSocketUrl }
+        }
+
+        override fun onMessage(webSocket: WebSocket, text: String?) {
+            if (text == null)
+                return
+
+            var type: String? = null
+            try {
+                val connect = Gson().fromJson(text, RtmConnect::class.java)
+                type = connect.type
+            } catch (e: Exception) {
+                e { "$e.message" }
+            }
+
+            if ("pong" != type) {
+                i { "Slack RTM message : $text" }
+            }
+
+            if (type != null) {
+                val eventListeners = listeners[type]
+                if (eventListeners != null && !eventListeners.isEmpty()) {
+                    for (listener in eventListeners) {
+                        listener.onMessage(text)
+                    }
+                }
+            }
+        }
+
+        override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
+            if (!closeListeners.isEmpty()) {
+                for (listener in closeListeners) {
+                    listener.onClose()
+                }
+            }
+        }
+
+        override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
+            if (t !is Exception)
+                return
+
+            if (!failureListeners.isEmpty()) {
```

Comment:
can be simplified

`                failureListeners.forEach{
                    it.onFailure(RuntimeException(t))
                }`